Fix Playoff Matches ending without a winner

Malaber opened the pull request.

Malaber committed:
Correct state of empty_prepared_playoff_match factory

Malaber committed:
Prevent matches from being stopped without a winner in playoffs

Malaber committed:
Split test methods for failed stopping of matches

Malaber committed:
Simplify controller code

Matches update now gets rolled back via a Transaction

Thor77 commented on app/controllers/matches_controller.rb:
```diff
@@ -13,22 +13,38 @@ def show
+
+    Match.transaction do
+      if @match.update(match_params)
+        return handle_match_end if new_state == 'finished'
```
what is the return doing here?

Malaber replied:
At some point in development it was necessary there.. removed it now

Thor77 commented on app/controllers/matches_controller.rb:
```diff
@@ -13,22 +13,38 @@ def show
+    return if @match.group_match?
+
+    if @match.winner.nil?
+      error_message = 'Stopping undecided Matches isn\'t allowed in playoff stage'
```
:thinking:

Malaber replied:
Line would be too long without it, but that can also be refactored, so done

Thor77 commented on app/controllers/matches_controller.rb:
```diff
@@ -13,22 +13,38 @@ def show
+    result = PopulateMatchBelowAndSave.call(match: @match)
+    return if result.success?
+
+    error_message = 'Moving Team one stage down failed'
```
:thinking:

Malaber replied:
Line would be too long without it, but it can also be refactored, so done

Thor77 commented on app/controllers/matches_controller.rb:
```diff
@@ -13,22 +13,38 @@ def show
+      raise ActiveRecord::Rollback
+    end
+
+    result = PopulateMatchBelowAndSave.call(match: @match)
```
:thinking:

Malaber replied:
removed

Thor77 commented on app/controllers/matches_controller.rb:
```diff
@@ -13,22 +13,38 @@ def show
+    end
+
+    result = PopulateMatchBelowAndSave.call(match: @match)
+    return if result.success?
```
i think a `unless result.success?` is easier to read here

Malaber replied:
if != unless :thinking:

Thor77 replied:
i'm aware. but i think a big unless block until the end of the method is easier to read

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -70,33 +70,69 @@
-          it 'updates the matches status' do
-            expect(response).to be_successful
-            expect(@running_playoff_match.state).to eq(finished[:state])
+            it 'updates the matches status' do
```
```suggestion
            it 'updates the match state' do
```

Malaber replied:
done

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -70,33 +70,69 @@
           end
 
-          describe 'updates the match below' do
+          context 'on a undecided match' do
```
```suggestion
          context 'on an undecided match' do
```

Malaber replied:
done

Thor77 commented on spec/controllers/matches_controller_spec.rb:
```diff
@@ -70,33 +70,69 @@
+              @running_playoff_match.reload
+            end
+
+            it 'returns unprocessable entity' do
```
```suggestion
            it 'returns an unprocessable entity response' do
```

Malaber replied:
done

Malaber committed:
Test returning unprocessable entity response

We now test sending a senseless state and also what happens when the
match.update method fails for some reason.

Malaber committed:
Test unprocessable entity response on a match that is not ready

Unprocessable entity is returned when you try to start a match that is
not ready yet.

Malaber committed:
Test returning unprocessable entity when error occurs in match population

Malaber committed:
Fix two typos

Thor77 reviewed (approved):
LGTM
